### PR TITLE
TGUI: Jurisdictional Automated System

### PR DESCRIPTION
**Pull request by Crowbar764**
# About the pull request

Converts the military police JAS terminal to TGUI.

<details>
<summary>Screenshots</summary>

### Main menu
image

### Populated report screen
image

### Add charge screen
image

### Import report screen
image

### Old UI for reference
image

</details>

**Question:** I made it so you cannot select the same charge twice. I am not an MP main, but as far as I can tell, duplicate charges have no impact on sentence time, so I have disabled the button if the charge is already added. Is this a mistake?

# Explain why it's good for the game

It's not. MPs should suffer and not have nice things.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

https://user-images.githubusercontent.com/59719612/209056758-2e6e0d01-ce0a-4821-8db5-2992759132ad.mp4

</details>

# Changelog

:cl: Casper
refactor: MP JAS computer now uses TGUI
/:cl:

### CODE REVIEW

**Comment by silencer-pl** [comment]
> Question: I made it so you cannot select the same charge twice. I am not an MP main, but as far as I can tell, duplicate charges have no impact on sentence time, so I have disabled the button if the charge is already added. Is this a mistake?

Yes, it is. Specifically in case of two charges - Minor and Major Disorderly Conduct in Brig, the charges from both variants of that one are meant to stack. 

Additionally, some MPs also use total counts on multiple charges, particularly on people who have done... A lot. 

The latter I can take or leave and it does not really influence anything other than the announcement, but the former would change how disorderly in brig works and is counted and would require a ML change, plus it would remove the additional detriment put in place against say wanton vandalism while in the brig.

File: code/game/machinery/computer/sentencing.dm [hunk 1]
```diff
@@ -1,481 +1,266 @@
...
+		ui.set_autoupdate(FALSE)
 
-		if(import(O))
-			ping("\The [src] pings, \"Successfully imported incident report!\"")
-			menu_screen = "incident_report"
-			updateUsrDialog()
-		else
-			to_chat(user, SPAN_ALERT("Could not import incident report."))
+/obj/structure/machinery/computer/sentencing/ui_data(mob/user)
+	. = list()
```
Comment (stanalbatross) [inline code/game/machinery/computer/sentencing.dm:26]: I will not force you to change this but typically we discourage `.`-hijacking for ui_data

File: code/game/machinery/computer/sentencing.dm [hunk 2]
```diff
@@ -1,481 +1,266 @@
...
-
-			if(error)
-				to_chat(usr, SPAN_ALERT(error))
+			var/datum/law/L = locate(params["charge"])
+			incident.charges -= L
+			incident.refresh_sentences()
+
+		if ("set_suspect")
+			if (!incident)
+				return TRUE
```
Comment (stanalbatross) [inline code/game/machinery/computer/sentencing.dm:130]: these checks should be performed in the UI not the DMcode

File: code/game/machinery/computer/sentencing.dm [hunk 3]
```diff
@@ -1,481 +1,266 @@
...
-				menu_screen = "main_menu"
-		if("print_imported")
-			var/error = print_incident_report(TRUE)
+				to_chat(usr, SPAN_INFO("You need the suspect ID in your hand, or grab them and use the terminal."))
+
+		if ("edit_summary")
+			if (!incident)
+				return TRUE
+
+			var/notes = strip_html((input(usr, "Describe the incident","Incident Report", html_decode(incident.notes)) as message), MAX_PAPER_MESSAGE_LEN)
```
Comment (stanalbatross) [inline code/game/machinery/computer/sentencing.dm:145]: tgui input text (big version) pls

File: tgui/packages/tgui/interfaces/Sentencing.js [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+                  icon="trash"
+                  onClick={() => act('remove_charge', { charge: charge.ref })}
+                />
+              </Flex.Item>
+            </Flex>
+          );
+        })}
+        <Flex justify="center" mt=".75rem">
+          <Button
+            content="new charge"
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/Sentencing.js:281]:
```suggestion
            content="New Charge"
```

File: tgui/packages/tgui/interfaces/Sentencing.js [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+  'new_charge': () => NewCharge,
+};
+
+export const Sentencing = (props, context) => {
+  const { data } = useBackend(context);
+  const { current_menu } = data;
+  const PageComponent = PAGES[current_menu]();
+
+  return (
+    <Window theme="weyland" width={780} height={725}>
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/Sentencing.js:18]: should be scrollable

File: tgui/packages/tgui/interfaces/Sentencing.js [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+  return (
+    <>
+      <Flex mb=".25rem" align="center">
+        <Button.Confirm
+          icon="arrow-left"
+          px="2rem"
+          textAlign="center"
+          mr="1rem"
+          onClick={() => act('set_menu', { new_menu: 'main' })}
+        />
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/Sentencing.js:121]: this seemingly sends the paper inside into nullspace and makes it inaccessible - you should properly get rid of the current incident / export it / explain it is being deleted when you do this

**Review by stanalbatross — CHANGES_REQUESTED** [review]

**Comment by Crowbar764** [comment]
Import page removed for simplicity sake. Main menu can still import reports:

image

File: tgui/packages/tgui/interfaces/Sentencing.js [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+          content="Export"
+          icon="print"
+          ml="auto"
+          px="2rem"
+          bold
+          tooltip={canExport ? '' : 'Missing suspect or charges'}
+          disabled={!canExport}
+          onClick={() => act('export')}
+        />
+      </Flex>
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/Sentencing.js:92]: I would definitely wrap this in a section tag for consistency, doesn't need a title

File: tgui/packages/tgui/interfaces/Sentencing.js [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+  return (
+    <>
+      <Flex mb=".25rem" align="center">
+        <Button.Confirm
+          icon="arrow-left"
+          px="2rem"
+          textAlign="center"
+          mr="1rem"
+          onClick={() => act('scrap_report')}
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/Sentencing.js:77]: needs to better communicate that this **will** scrap and delete the report

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/Sentencing.js (hunk 8 above)
Comment (Crowbar764) [reply tgui/packages/tgui/interfaces/Sentencing.js]: Added a tooltip, is this enough?

Comment (stanalbatross) [reply tgui/packages/tgui/interfaces/Sentencing.js]: sure

**Review by stanalbatross — APPROVED** [review]

**Pull request merged**